drivers: i2c: i2c_eos_s3: Add I2C driver support for EOS S3

Willmish opened the pull request:
# Module dependent PR
> **Warning**
> This PR depends on https://github.com/zephyrproject-rtos/hal_quicklogic being updated
 **DO NOT** merge this PR until https://github.com/zephyrproject-rtos/hal_quicklogic/pull/4 is merged

## Description

This PR adds basic i2c driver implementation for the EOS S3 SoC, used with the [Sparkfun QuickLogic Thing Plus](https://github.com/sparkfun/QuickLogic_Thing_Plus) board (which is very similar to the intree [QuickLogic QuickFeather](https://docs.zephyrproject.org/latest/boards/arm/quick_feather/doc/index.html) ) based on the HAL module ported from [QuickLogic's QORC SDK](https://github.com/QuickLogic-Corp/qorc-sdk/).

Since there is an API mismatch between the actual HAL and the zephyr i2c driver API, I left a comment explaining my current approach in `drivers/i2c/i2c_eos_s3.c` file (https://github.com/zephyrproject-rtos/zephyr/compare/main...Willmish:zephyr:eos_s3_fixes?expand=1#diff-586321414cf511e0b1f1b99e041b736fe987a5d03a43eb9073e47c7ca2943f26R152). I would strongly appreciate guidance whether this is a correct approach or not.

Additionally this PR adds a tiny change to set up pinmuxing for the builtin button pad by default (this is normally muxed by the default bootloader provided by QuickLogic, but IMO the board port should be bootloader independent, so planning to add future PRs for elements of the bootloader that are taken for granted/add support for MCUBoot)
 
## Comprehensive list of changes:
- [ ] Update Kconfig.defconfig to include a BUILD_OUTPUT_BIN boolean flag - not sure if this is absolutely necessary, but current flasher for all EOS S3 based boards - [TinyFPGAProgrammer](https://github.com/QuickLogic-Corp/TinyFPGA-Programmer-Application)  requires  a binary file for flashing
- [ ] Add Button pad pinmuxing to `board.c` file for quickfeather board
- [ ] add necessary defines to  `board.h` and `soc_pinmap.h` files for the abovementioned pinmuxing for the `quickfeather` board
- [ ] Add all required files for i2c driver, including:
  - [ ] Updated `CMakeLists.txt`
  - [ ] Updated `Kconfig`
  - [ ] Added `Kconfig.eos_s3`
  - [ ] Added `i2c_eos_s3.c`
  - [ ] Added `quicklogic,eos-s3-i2c.yaml` dts binding
- [ ] Additionally : Updated `soc/arm/quicklogic_eos_s3/soc.c` to enable Flexible Fusion Engine Clocks, setup its dividors and properly gate/route them to be used by I2C peripherals (based on [EOS S3 Technical Reference Manual, Chapter 36.1 Setup FFE clocks, page 340](https://www.quicklogic.com/wp-content/uploads/2020/06/QL-S3-Technical-Reference-Manual-Vol-1.pdf)
- [ ] Updated SoC .dtsi file to `quicklogic_eos_s3.dtsi` to define the i2c0 device (there are two, should I also add a skeleton definition for i2c1?)
- [ ] Updated QuickFeather board dts to enable i2c0 by default

## Additional info
Opening as draft for now to be visible in https://github.com/zephyrproject-rtos/hal_quicklogic/pull/4 , but requires commit cleanup and code cleanup to conform to [Zephyr's Contribution guidelines](https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html)

Created in collaboration with @JDuchniewicz in preparation for [Embedded Open Source Summit 2023](https://eoss2023.sched.com/event/1LcM4/porting-an-ai-powered-wearable-health-monitor-to-zephyr-on-open-hardware-szymon-duchniewicz-avanade-jakub-duchniewicz-tietoevry)

JDuchniewicz committed:
drivers: i2c: eos_s3: First preparation for I2C for EOS S3 in tree

Adds initial modifications to CMakeLists.txt file, Kconfig file,
creates a Kconfig.eos_s3 file and introduces a skeleton code for the
driver based on drivers/i2c/i2c_sifive.c.

Signed-off-by: Jakub Duchniewicz <j.duchniewicz@gmail.com>

Willmish committed:
boards: arm: quick_feather: Add button pinmux.

Modifies board initialisation file to setup proper pinmuxing for the
onboard user-programmable button. (Currently this was assumed to be
handled in the bootloader) Also modifies
soc/arm/quicklogic_eos_s3/soc_pinmap.h to follow the same macro format
as uart pinmux.

Signed-off-by: Szymon Duchniewicz <szduchniewicz@gmail.com>

JDuchniewicz committed:
dts: arm: quicklogic: Add I2C DTS definition for  EOS S3.

Add initial I2C definition to the quicklogic_eos_s3.dtsi Device Tree
overlay. Enable i2c0 for quick_feather board.

Signed-off-by: Jakub Duchniewicz <j.duchniewicz@gmail.com>

Willmish committed:
boards: arm: quick_feather: Kconfing.defconfig and dts update.

Update quick_feather Kconfig.defconfig to add BUILD_OUTPUT_BIN config
option. Update quick_feather dts to add copyright attribution.

Signed-off-by: Szymon Duchniewicz <szduchniewicz@gmail.com>

henrikbrixandersen commented:
> > Warning
> > This PR depends on https://github.com/zephyrproject-rtos/hal_quicklogic being updated
> > **DO NOT** merge this PR until [zephyrproject-rtos/hal_quicklogic#4](https://github.com/zephyrproject-rtos/hal_quicklogic/pull/4) is merged

In order to get this right, this PR needs to pull in https://github.com/zephyrproject-rtos/hal_quicklogic/pull/4 using `west.yml`, as we need a CI run containing the _combined_ the changes before merging https://github.com/zephyrproject-rtos/hal_quicklogic/pull/4

Willmish committed:
west.yml: bump hal_quicklogic module

This commit changes module revision to point to the pull request adding
required hal changes.

Signed-off-by: Szymon Duchniewicz <szduchniewicz@gmail.com>

Willmish commented:
> 
> In order to get this right, this PR needs to pull in [zephyrproject-rtos/hal_quicklogic#4](https://github.com/zephyrproject-rtos/hal_quicklogic/pull/4) using `west.yml`, as we need a CI run containing the _combined_ the changes before merging [zephyrproject-rtos/hal_quicklogic#4](https://github.com/zephyrproject-rtos/hal_quicklogic/pull/4)

Ah! I was wondering if I can somehow add this module to CI, thanks for the pointer @henrikbrixandersen . Since my fork is sitting outside of the zephyrproject org, I modified the SHA revision and changed pointer to repo from `repo-path` to `url` pointing to my fork. This is done in: https://github.com/zephyrproject-rtos/zephyr/pull/59905/commits/9509d0e435cbfc59a21613a6a0ea682e61f1f9aa

Q1: Is that correct? and then after merging https://github.com/zephyrproject-rtos/hal_quicklogic/pull/4 but before merging this, I'd change the commit to point to the tip commit SHA and revert the repo reference back to `repo-path` for the module?

Q2: I remember us talking during EOSS about rewriting the driver to directly do memory writes for the WishBone bus, rather than using a HAL. My current plan was to first conform with the QuickLogic HAL, and then when feeling confident enough dealing with Zephyr drivers and the SoC itself, update it to remove that abstraction layer. I am still unsure how to abstract out the common Wishbone bus initialisation, transmit and receive functions (https://github.com/Willmish/hal_quicklogic/blob/hal_i2c/HAL/src/eoss3_hal_wb.c) so that they would be accessible by all drivers (I2C, SPI etc), rather than maintaining copies of it for each driver. Possibly this is what would be kept in the module?
 Does that sound like a good roadmap plan?

fkokosinski commented on west.yml:
```diff
@@ -193,9 +193,9 @@ manifest:
         - hal
     - name: hal_quicklogic
-      revision: b3a66fe6d04d87fd1533a5c8de51d0599fcd08d0
+      revision: c12e7ccb2da626c53f264e7109ff51324fe4c9da
```
Please change `revision` to `pull/4/head` so that it references your PR to the QL HAL, and drop `repo-path`/`url` changes.

Willmish commented:
What would be the best approach to conform to the coding standards? should I apply a fix patch or squash most commits and apply the appropriate coding styles to each squashed commit in history? Looking at: https://docs.zephyrproject.org/latest/contribute/guidelines.html#clang-format

Willmish committed:
soc: arm: eos_s3: Add FFE clock enabling

* Add Flexible Fusion Engine Clock enabling for Wishbone bus communication
and I2C communication.

Signed-off-by: Szymon Duchniewicz <szduchniewicz@gmail.com>

Willmish committed:
drivers: i2c: eos_s3 i2c driver fixes

* Add dts bindings for i2c driver for EOS S3.
* Add fixs to EOS S3 dts/dtsi files.
* Fix wrong Kconfig Macro - not matching the one generated from
dts/bbinding/quicklogic,eos-s3-i2c.yaml.
* drivers: i2c: i2c_eos_s3.c: Fix compilation errors.

Signed-off-by: Szymon Duchniewicz <szduchniewicz@gmail.com>

Willmish commented:
Cool, thanks @fkokosinski . Squashed some commits, cleaned up a bit and fixed commit messages, so should be ready for review? I may need some guidance in properly coding this function (as you might be able to tell by the comments): https://github.com/zephyrproject-rtos/zephyr/pull/59905/files#diff-586321414cf511e0b1f1b99e041b736fe987a5d03a43eb9073e47c7ca2943f26R131

But this should be passing all compliance checks and tests

fkokosinski commented:
> Hi, could I request the stale label to be removed so the pr doesnt close out? I was out for some of the summer so was unable to finish this off, but planning to do so in the next few weeks

Sure thing. :) Done
